src/preprocessing/passes/non_clausal_simp.cpp

```diff
@@ -448,10 +444,105 @@ PreprocessingPassResult NonClausalSimp::applyInternal(
         Rewriter::rewrite(Node(learnedBuilder)));
   }
 
-  propagator->setNeedsFinish(true);
   return PreprocessingPassResult::NO_CONFLICT;
 }  // namespace passes
 
+std::pair<bool, std::vector<Node>> NonClausalSimp::preprocessByCryptominisat(
+    AssertionPipeline* assertionsToPreprocess, unsigned substs_index)
+{
+  std::unique_ptr<SatSolver> d_satSolver(SatSolverFactory::createCryptoMinisat(
+      smtStatisticsRegistry(), "non_clausal_simp_solver"));
+  CVC4::prop::NullRegistrar d_registrar;
+  context::Context d_context;
+  CVC4::prop::TseitinCnfStream d_cnfStream(d_satSolver.get(),
+                                           &d_registrar,
+                                           &d_context,
+                                           true,
+                                           "toCNF-skeleton-preprocessing");
+  std::vector<Node> learned_literals;
+  Trace("skeleton-preprocessing") << "asserting to sat solver" << std::endl;
+  {  // for timing
+    TimerStat::CodeTimer timer(d_statistics.d_addingAssertions);
+    for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+    {
+      Assert(Rewriter::rewrite((*assertionsToPreprocess)[i])
+             == (*assertionsToPreprocess)[i]);
+      // Don't reprocess substitutions
+      if (substs_index > 0 && i == substs_index)
+      {
+        continue;
+      }
+      Trace("skeleton-preprocessing")
+          << "asserting " << (*assertionsToPreprocess)[i] << std::endl;
+      d_cnfStream.convertAndAssert(
+          (*assertionsToPreprocess)[i], false, false, RULE_GIVEN);
+    }
+  } // end timing
+  Trace("non-clausal-simplify") << "solving" << std::endl;
+  if (d_satSolver->solve() == SAT_VALUE_FALSE)
+  {
+    // If in conflict, just return false
+    return std::make_pair(false, learned_literals);
+  }
+
+  std::vector<SatLiteral> topLevelUnits = d_satSolver->getTopLevelUnits();
+  // we start from index 2 because the first two top level units much be True
+  // and Not False
+  for (size_t i = 2; i < topLevelUnits.size(); i++)
+  {
+    SatLiteral lit = topLevelUnits[i];
+    Node learnedLiteral = d_cnfStream.getNode(lit);
```

Comment:
Is it guaranteed that all literals returned by CryptoMiniSat have a corresponding node in CVC4?  CryptoMiniSat supports bounded variable addition as a simplification technique, which adds new variables. CryptoMiniSat may potentially return literals that we don't have mapped to nodes.